File: Striver/Trees/Q16.py
# boundary traversal of binary tree - anti-clockwise
from Node import Node

def leftTraversal(root, res):
    if not root:
        return
    if root.left:
        res.append(root.data)
        leftTraversal(root.left, res)
    elif root.right:
        res.append(root.data)
        leftTraversal(root.right, res)

def rightTraversal(root, res):
    if not root:
        return
    if root.right:
        rightTraversal(root.right, res)
        res.append(root.data)
    elif root.left:
        rightTraversal(root.left, res)
        res.append(root.data)

def leafTraversal(root, res):
    if not root:
        return
    leafTraversal(root.left, res)
    if not root.left and not root.right:
        res.append(root.data)
    leafTraversal(root.right, res)

def boundaryTraversal(root):
    if not root:
        return []
    res = []
    res.append(root.data)
    leftTraversal(root.left, res)
    leafTraversal(root, res)
    rightTraversal(root.right, res)
    return res

Node1 = Node(20)
Node1.left = Node(8)
Node1.right = Node(22)
Node1.left.left = Node(4)
Node1.left.right = Node(12)
Node1.left.right.left = Node(10)
Node1.left.right.right = Node(14)
Node1.right.right = Node(25)
print(boundaryTraversal(Node1))

File: Striver/Trees/Q16.java
package Striver.Trees;
import java.util.*;

public class Q16 {
  // boundary traversal of a binary tree - anticlockwise
  
  public static void LeftTraversal(Node root, ArrayList<Integer> ans){
      if(root == null) return;
      if(root.left != null){
          ans.add(root.data);
          LeftTraversal(root.left, ans);
      }
      else if(root.right != null){
          ans.add(root.data);
          LeftTraversal(root.right, ans);
      }
  }

    public static void RightTraversal(Node root, ArrayList<Integer> ans){
        if(root == null) return;
        if(root.right != null){
            RightTraversal(root.right, ans);
            ans.add(root.data);
        }
        else if(root.left != null){
            RightTraversal(root.left, ans);
            ans.add(root.data);
        }
    }

    public static void LeafTraversal(Node root, ArrayList<Integer> ans){
        if(root == null) return;
        LeafTraversal(root.left, ans);
        if(root.left == null && root.right == null) ans.add(root.data);
        LeafTraversal(root.right, ans);
    }

    public static ArrayList <Integer> printBoundary(Node root)
    {
        ArrayList<Integer> ans = new ArrayList<>();
        if(root == null) return ans;
        ans.add(root.data);
        LeftTraversal(root.left, ans);
        LeafTraversal(root, ans);
        RightTraversal(root.right, ans);
        return ans;
    }

    public static void main(String[] args) {
        Node root = new Node(20);
        root.left = new Node(8);
        root.left.left = new Node(4);
        root.left.right = new Node(12);
        root.left.right.left = new Node(10);
        root.left.right.right = new Node(14);
        root.right = new Node(22);
        root.right.right = new Node(25);
        ArrayList<Integer> ans = printBoundary(root);
        for(int i : ans) System.out.print(i + " ");
    }
}
